pimd/pim_cmd.c

```diff
@@ -3600,17 +3610,74 @@ static void igmp_show_sources(struct pim_instance *pim, struct vty *vty)
 				pim_time_uptime(uptime, sizeof(uptime),
 						now - src->source_creation);
 
-				vty_out(vty, "%-16s %-15s %-15s %5s %3s %8s\n",
-					ifp->name, group_str, source_str, mmss,
-					IGMP_SOURCE_TEST_FORWARDING(
-						src->source_flags)
-						? "Y"
-						: "N",
-					uptime);
+				if (uj) {
+					json_object_object_get_ex(
+						json, ifp->name, &json_iface);
+					if (!json_iface) {
+						json_iface =
+							json_object_new_object();
+						json_object_string_add(
+							json_iface, "name",
+							ifp->name);
+						json_object_object_add(
+							json, ifp->name,
+							json_iface);
+					}
+					json_object_object_get_ex(json_iface,
+								  group_str,
+								  &json_group);
+
+					if (!json_group) {
+						json_group =
+							json_object_new_object();
+						json_object_string_add(
+							json_group, "group",
+							group_str);
+						json_object_object_add(
+							json_iface, group_str,
+							json_group);
+						json_sources =
+							json_object_new_array();
+						json_object_object_add(
+							json_group, "sources",
+							json_sources);
+					}
+					json_source = json_object_new_object();
+					json_object_string_add(json_source,
+							       "source",
+							       source_str);
+					json_object_string_add(json_source,
+							       "timer", mmss);
+					json_object_boolean_add(
+						json_source, "forwarded",
+						IGMP_SOURCE_TEST_FORWARDING(
+							src->source_flags));
+					json_object_string_add(
+						json_source, "upTime", uptime);
```

Comment:
I thought `uptime` is a single word :) Should be maybe just `uptime` and not `upTime`?